src/cargo/util/errors.rs

```diff
@@ -300,6 +302,140 @@ impl From<clap::Error> for CliError {
     }
 }
 
+/// Error that comes from running `git` CLI. This wrapper exists to detect
+/// this kind of errors in [`crate::util::network::with_retry`] and properly
+/// retry them.
+pub struct GitCliError {
+    inner: anyhow::Error,
+}
+
+impl GitCliError {
+    /// Creates an insteance of [`GitCliError`].
+    ///
+    /// # Invariants
+    ///
+    /// The inner error must by constructed from [`ProcessError`] that
+    /// was returned as a result of running `git` CLI operation.
+    /// If it isn't true, the code won't panic, but [`Self::is_retryable()`]
+    /// will always return `false`.
+    pub fn new(inner: anyhow::Error) -> Self {
+        Self { inner }
+    }
+
+    /// Returns `true` if the git CLI error is transient and may be retried.
+    ///
+    /// The code here is inspired by `chromium` git retries code:
+    /// <https://chromium.googlesource.com/infra/infra/+/b9f6e35d2aa1ce9fcb385e414866e0b1635a6c77/go/src/infra/tools/git/retry_regexp.go#17>
+    pub fn is_retryable(&self) -> bool {
+        const GIT_TRANSIENT_ERRORS: &[&str] = &[
+            // Error patterns are taken from chromium.
+            // Please, keep them 1-to-1 exactly the same as in their source code
```

Comment:
I don't think that we need to be so strict here. For example I don't think that we'll be looking into their source code every-so-often to see if these lists match, and it's not like someone should send a PR to chromium before updating the list here.